src/cocoa/toga_cocoa/widgets/tree.py

```diff
@@ -203,24 +235,31 @@ def create(self):
         # Add the layout constraints
         self.add_constraints()
 
-    # def data_changed(self, node=None):
-    #     if node:
-    #         if node._expanded:
-    #             self.tree.expandItem(node._impl)
-    #         else:
-    #             self.tree.collapseItem(node._impl)
-
     def change_source(self, source):
         self.tree.reloadData()
 
     def insert(self, parent, index, item):
-        self.tree.reloadData()
+
+        # set parent = None if inserting to the root item
+        index_set = NSIndexSet.indexSetWithIndex(index)
+        if parent is self.interface.data:
+            parent = None
+        else:
+            parent = getattr(parent, '_impl', None)
+
+        self.tree.insertItemsAtIndexes(index_set, inParent=parent, withAnimation=NSTableViewAnimation.SlideDown.value)
 
     def change(self, item):
-        self.tree.reloadData()
+        if hasattr(item, '_impl'):  # item is or was visible -> NSOutlineView knows it
```

Comment:
It is *very* slightly faster to use a `try: except AttributeError` - that way, there's only ever one attempt to access the attribute, and exceptions are faster than a comparison.

Reply:
Changed.